Add in-game updater for linux platforms

Diyou opened the pull request:
- Addresses #14653

Added experimental Update Support for the AppImage Client (Linux)

In order for this to work [appimageupdatetool](https://github.com/AppImage/AppImageUpdate) needs to be installed either systemwide or shipped within the osu.AppImage.
They have release versions in AppImage format but building the executable and library might be more preferable: [appimageupdate.tar.gz](https://github.com/ppy/osu/files/9103098/appimageupdate.tar.gz)

I'm not very experienced with the osu framework so there are still a lot of things that need to be cleaned up for proper functionality:

- Get the proper path to the osu.Appimage ( arg0 ?)
- Restart the App when clicking on the UpdateCompleteNotification
- Adjust the StartInfo.FileName to match the shipped binary

Diyou committed:
Added: AppImageUpdateManager

Added experimental Update Support for the AppImage Client (Linux)

peppy commented:
This will require some serious massaging.

- Lots of duplicated code
- Code formatting rules not applied
- The actual work code is nested inside a delegate. Move it to its own method so it's at least legible.
- UX is completely different to squirrel version for no good reason?

Diyou committed:
Escaped quotes to support paths with spaces

Diyou committed:
Update AppImageUpdateManager.cs

Diyou committed:
Added nested class AppImageUpdateTool

Moved the working code responsible for updating into a designated class with extended functionality.

Diyou committed:
Update AppImageUpdateManager.cs

Diyou commented:
- Moved the working code into a class which also manages its progress states and other convenient attributes and methods.
- If no appimageupdatetool is installed (either systemwide or within the appimage) it falls back to SimpleUpdateManager

TODOs: 
- Implement proper restart functionality for the ProgressCompletionNotification.
- I could not test whether the environment variable APPIMAGE gets set via the AppImage.
```csharp
ProcessStartInfo info = new ProcessStartInfo
{
    FileName = Command,
    //TODO For Debugging put a test osu.AppImage next to the osu binary
    Arguments = $"{arguments}{argument_spacer}\"{RuntimeInfo.StartupDirectory}/osu.AppImage\"",
    //TODO For deployed builds use the APPIMAGE environment variable
    //Arguments = $"{arguments}{argument_spacer}\"{ImagePath}\"",
    UseShellExecute = false,
    RedirectStandardError = true,
    RedirectStandardOutput = true,
};
```
- Create an Interface to reduce duplicated code and normalised UX?

Diyou committed:
Update AppImageUpdateManager.cs

Diyou committed:
Update AppImageUpdateManager.cs

Diyou committed:
Added Restart functionality to AppImageUpdateManager

Note: Closes the App instantly ... should say bye atleast

Diyou committed:
Removed empty lines

Diyou committed:
Update AppImageUpdateManager.cs

Diyou committed:
Added some doc & minor consistency improvements

Diyou committed:
Fixed UX for applying update(restarting the game) and improved legibility(+docs)

Everything outside AppImageUpdateTool looks and should behave completely like the SquirrelUpdateManager

Diyou commented:
I cleaned out the toolset specific code in the UpdateManager and now looks closely like the Squirrel version.
Maybe create a platform agnostic template?

frenzibyte commented:
Please refrain from opening PRs off your master branch, as that makes it hard for us to push changes to the PR for merge. Have a read on the [contributing guidelines](https://github.com/ppy/osu/blob/279bdcb3a43a70af50ff2f103b270fbe3dc2d81b/CONTRIBUTING.md).

frenzibyte commented on osu.Desktop/Updater/AppImageUpdateManager.cs:
```diff
@@ -0,0 +1,564 @@
+    [SupportedOSPlatform("linux")]
+    public class AppImageUpdateManager : UpdateManager
+    {
+        /// <summary>
+        /// Implements appimageupdatetool functionality via cli
+        /// </summary>
+        private class AppImageUpdateTool
```
Move this to a different file as a `partial class` of this file.

bdach replied:
I don't know about partials. This should probably live as a separate top-levelc lass in `osu.Desktop/Linux` (which doesn't currently exist yet).

frenzibyte commented on osu.Desktop/Updater/AppImageUpdateManager.cs:
```diff
@@ -0,0 +1,564 @@
+            /// <item><see cref="TOOL_HELP"/></item>
+            /// </list>
+            /// </summary>
+            public enum ARGS
```
Keep enum names explanatory and in pascal casing, something like `ToolArguments`. Same applies to all other enums and enum members as well.

frenzibyte commented on osu.Desktop/Updater/AppImageUpdateManager.cs:
```diff
@@ -0,0 +1,564 @@
+            }
+            /// TODO: Set commandFile to match the shipped binary to prevent searching and remove the unnecessary code
+            private string commandFile;
+            private string[] lookFor =
```
Also use explanatory names here, e.g. `toolLookupFilenames`.

frenzibyte commented on osu.Desktop/Updater/AppImageUpdateManager.cs:
```diff
@@ -0,0 +1,564 @@
+                COMPLETED,
+                CANCELLED
+            }
+            private static AppImageUpdateTool instance;
+            public static AppImageUpdateTool Instance
+            {
+                get
+                {
+                    if (instance is default(AppImageUpdateTool))
+                    {
+                        instance = new AppImageUpdateTool();
+                    }
+                    return instance;
+                }
+            }
```
Is there a reason for the staticness in this class? Why not just always keep it based on instance and instantiate the tool in `AppImageUpdateManager`.

Diyou replied:
I think the only reason for the class to be static is to check its availability( AppImageUpdateTool.IsInstalled ) before instantiating the UpdateManager so it can fallback to SimpleUpdateManager.

frenzibyte commented on osu.Desktop/Updater/AppImageUpdateManager.cs:
```diff
@@ -0,0 +1,564 @@
+            {
+                get
+                {
+                    return Environment.GetEnvironmentVariable("APPIMAGE") is string;
```
Use pattern-matching when the returned string is used (same applies to all usages of this pattern).
```suggestion
                    return Environment.GetEnvironmentVariable("APPIMAGE") != null;
```

frenzibyte commented on osu.Desktop/Updater/AppImageUpdateManager.cs:
```diff
@@ -0,0 +1,564 @@
+                        process.Start();
+
+                        process.WaitForExit();
+                        return process.ExitCode == 1;
```
The tool finds updates if it returns... 1? Isn't that a failure code?

Diyou replied:
[wiki](https://github.com/AppImage/AppImageUpdate/wiki/Self-updating-AppImages#using-the-existing-appimageupdate-user-interfaces-to-perform-updates)
> Both AppImageUpdate and appimageupdatetool provide an easy to use update check. Just call those apps with the -j flag, and watch the exit code. If it exits with return value 0, the update check succeeded, and there is no newer version available. Status 1 means that the update check succeeded, and an update is available. Any other return code can be considered an error while trying to check for changes.

The tool also writes to STDERR when you using the help or version flag and some messages during updating which I did nor parse.

frenzibyte commented on osu.Desktop/Updater/AppImageUpdateManager.cs:
```diff
@@ -0,0 +1,564 @@
+                    {
+                        notificationOverlay.Post(new SimpleNotification
+                        {
+                            Text = $"A newer release of osu! has been found ({currentTagName} → {latestTagName}).\n\n"
+                                    + "Click here to download the new version.",
```
Probably want update to be initiated automatically, similar to Squirrel on Windows.

Diyou replied:
While doable I was reading in the [doc](https://docs.appimage.org/packaging-guide/optional/updates.html#recommended-user-experience) ?

> Recommended user experience[](https://docs.appimage.org/packaging-guide/optional/updates.html#recommended-user-experience)
> One advantage of the AppImage format is that it gives full control to application authors over the end user experience. Hence, using AppImage and AppImageUpdate, application authors can implement almost any of the schemes outlined above. In order to maintain a consistent and positive user experience with AppImages and AppImageUpdate, we recommend application authors to follow the following Golden Rules:
> 
> Never download updates without the user’s explicit consent, be it in the form of per-update consent, or, optionally, opt-in consent for automatic updates. Thanks for not killing users’ mobile data plans by downloading stuff without asking
> 
> Respect global flags for “do not check for new versions” and “do not attempt to update”. The user may be running a central updating daemon that manages updates for the whole system, in which case any and all attempts to update the application from within itself should be skipped. We need to define those flags for 1) per-system and 2) per-user configuration and 3) ENV (similar to how the old desktopintegration script was set up not to interfere with appimaged)
> 
> Do not bother the user with updates directly as the first thing when the application is launched. When opening an application for the first time, users should see something meaningful to give a positive impression and show immediately what the application is all about (after all, we are automatically taking a screenshot of what your application shows directly after it has been launched for AppImageHub)
> 
> Ask the user for permission before doing version checks. Many open source users value privacy highly and don’t appreciate the “phone home” aspect of forced version checks, which effectively are a form of tracking

Its a general thing though and thus a different issue.

frenzibyte replied:
Well, if that's how we have it working on Windows I'm not sure Linux should be any different? cc @peppy

frenzibyte commented on osu.Desktop/Updater/AppImageUpdateManager.cs:
```diff
@@ -0,0 +1,564 @@
+
+            protected override Notification CreateCompletionNotification()
+            {
+                return new UpdateCompleteNotification(updateManager);
+            }
+
+            [BackgroundDependencyLoader]
+            private void load(OsuColour colours)
+            {
+                IconContent.AddRange(new Drawable[]
+                {
+                    new Box
+                    {
+                        RelativeSizeAxes = Axes.Both,
+                        Colour = ColourInfo.GradientVertical(colours.YellowDark, colours.Yellow)
+                    },
+                    new SpriteIcon
+                    {
+                        Anchor = Anchor.Centre,
+                        Origin = Anchor.Centre,
+                        Icon = FontAwesome.Solid.Upload,
+                        Colour = Color4.White,
+                        Size = new Vector2(20),
+                    }
+                });
+            }
+
+            public override void Close()
+            {
+                // cancelling updates is not currently supported by the underlying updater.
+                // only allow dismissing for now.
+
+                switch (State)
+                {
+                    case ProgressNotificationState.Cancelled:
+                        base.Close();
+                        break;
+                }
+            }
+        }
```
Probably want these shared between here and `SquirrelUpdateManager`, and move methods like `PrepareUpdateAsync` which are used here in the notification to the base `UpdateManager` class as a virtual and override in both managers.

frenzibyte reviewed (changes requested):
Initial / code-quality pass, not necessarily a linux user. You may want to have a look at running code inspection and fix code style, see `InspectCode.sh`.

Diyou committed:
Applied CodeAnalysis

Diyou commented:
Was using vscode without running automatic Code Analysis (sorry about that).
You might wanna add a settings.json to the .vscode folder and add
```json
{
    "omnisharp.enableRoslynAnalyzers": true
}
```

frenzibyte commented:
> Was using vscode without running automatic Code Analysis (sorry about that). You might wanna add a settings.json to the .vscode folder and add
> 
> ```json
> {
>     "omnisharp.enableRoslynAnalyzers": true
> }
> ```

Feel free to add that in a separate PR as long as it works properly.

bdach commented on osu.Desktop/Updater/AppImageUpdateManager.cs:
```diff
@@ -0,0 +1,551 @@
+    {
+        /// <summary>
+        /// Arguments passed to the appimageupdatetool
+        /// <list type="bullet">
+        /// <item><see cref="Check"/><description></description></item>
+        /// <item><see cref="Update"/></item>
+        /// <item><see cref="Overwrite"/></item>
+        /// <item><see cref="ToolVersion"/></item>
+        /// <item><see cref="ToolHelp"/></item>
+        /// </list>
```
Please remove this useless xmldoc. It basically serves no purpose.

bdach commented on osu.Desktop/Updater/AppImageUpdateManager.cs:
```diff
@@ -0,0 +1,551 @@
+        {
+            get
+            {
+                if (commandFile is default(string))
+                {
+                    foreach (string file in toollookupfilenames)
+                    {
+                        commandFile = file;
+                        if (IsInstalled) break;
+                    }
+                }
```
The way this is written is so wrong to me. This needs to be rewritten basically from scratch.

* `commandFile is default(string)` is probably one of the most obfuscated ways to write a null check that I've seen. And `commandFile` should be a `string?` to begin with (the expectation is that all new code should be nullable-annotated). This goes everywhere else, basically every usage of `is default(T)` should be removed.
* `IsInstalled` is shocking. It's a static property, which is derived from the value of `Version`, which is a property that *internally fires a process* via `getOutput()`. This violates all sorts of expectations that reside in idiomatic C# code. Properties should be for simple data access. If you want to fire processes, then please use normal methods to indicate that there may be side effects better. And as mentioned above, please get rid of all of the staticness. There's no real reason for it.

bdach commented on osu.Desktop/Updater/AppImageUpdateManager.cs:
```diff
@@ -0,0 +1,551 @@
+        private static ProcessStartInfo startInfo(ToolArguments args = ToolArguments.Check)
+        {
+            string arguments = "";
+            string argumentSpacer = " ";
```
This variable shouldn't even exist. Inline it at the only usage site.

bdach commented on osu.Desktop/Updater/AppImageUpdateManager.cs:
```diff
@@ -0,0 +1,551 @@
+
+        public enum States
+        {
+            NOUPDATESAVAILABLE,
+            UPDATESAVAILABLE,
+            DOWNLOADING,
+            VERIFYING,
+            COMPLETED,
+            CANCELLED
```
should use `CamelCase` rather than all caps

bdach commented on osu.Desktop/Updater/AppImageUpdateManager.cs:
```diff
@@ -0,0 +1,551 @@
+            }
+            catch
+            {
+                return false;
```
exceptions suppressed with no error logging whatsoever?

bdach commented on osu.Desktop/Updater/AppImageUpdateManager.cs:
```diff
@@ -0,0 +1,551 @@
+        {
+            get
+            {
+                if (version is default(string))
+                {
+                    string output = getOutput(ToolArguments.ToolVersion);
+
+                    if (output != null)
+                    {
+                        var match = Regex.Match(output, @"version\s([a-zA-Z0-9\._-]*).*$", RegexOptions.IgnoreCase);
+
+                        if (match.Success)
+                        {
+                            output = match.Groups[1].ToString();
+                        }
+                    }
+
+                    version = output;
+                }
```
code block needs to have less nesting than this `if { ... if { ... if {` structure. invert the checks and early-return instead to have it flatter

bdach commented on osu.Desktop/Updater/AppImageUpdateManager.cs:
```diff
@@ -0,0 +1,551 @@
+
+                                if (match.Groups[3].Success && match.Groups[4].Success)
+                                {
+                                    //float downloadedSize = float.Parse(match.Groups[3].ToString());
+                                    //float downloadSize = float.Parse(match.Groups[4].ToString());
```
commented code should be removed

bdach reviewed (changes requested):
Sorry but I don't even know where to begin with this diff, and describing how to fix it would likely take inordinately more time than rewriting it myself from scratch. The way things are done here is just not acceptable to me.

frenzibyte commented:
Closing as this has been PR'd from a side branch in #19167 rather than `master`.
